### PR TITLE
Make `Evapotranspiration` differentiable and vectorized

### LINKED ISSUE #44: [Task]: Make `Evapotranspiration` differentiable and efficient
the module [`Evapotranspiration`](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/evapotranspiration.py) has several classes. We skip `Simple_Evapotranspiration` because it is not used in wofost models. See explanations and requirements in https://github.com/WUR-AI/diffWOFOST/issues/39.

### PR DESCRIPTION
**Pull request by SCiarella**
Closes #44.

### CODE REVIEW

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+    # is non-differentiable at `DEPNR==3` and causes numerical vs autograd
+    # gradient mismatches when treating DEPNR as a continuous tensor.
+    #
+    # To keep regression behaviour intact we preserve exact values at the
+    # discrete DEPNR values used in the YAML fixtures (2.0/3.0/3.5/4.5):
+    # - DEPNR <= 2: full correction
+    # - DEPNR >= 3: no correction
+    # and smoothly transition (C1) between 2 and 3 using a cubic smoothstep.
+    t = DEPNR - 2.0
+    s = 3.0 * t**2 - 2.0 * t**3  # smoothstep on [0,1]
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]: I couldn't understand why we use "cubic smoothstep" as an approximation and how the coefficients are derived. I see that the problem here is somewhat similar to [SPAN variable in leaf_dynamics](https://github.com/WUR-AI/diffWOFOST/blob/c5318f7f4e32c44be3afc4b3a5bd21ddf06d1487/src/diffwofost/physical_models/crop/leaf_dynamics.py#L302-L323). Can we instead use STE method with sigmoid approximation here?

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+        shape: tuple | None = None,
+    ) -> None:
+        """Select and initialize the evapotranspiration implementation.
+
+        Chooses between layered CO2-aware, non-layered CO2, or standard evapotranspiration
+        based on available parameters.
+        """
+        if "soil_profile" in parvalues:
+            self.etmodule = EvapotranspirationCO2Layered(day, kiosk, parvalues, shape=shape)
+        elif "CO2TRATB" in parvalues:
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
        elif "CO2" in parvalues and "CO2TRATB" in parvalues:
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+            self.etmodule = EvapotranspirationCO2(day, kiosk, parvalues, shape=shape)
+        else:
+            self.etmodule = Evapotranspiration(day, kiosk, parvalues, shape=shape)
+
+    @prepare_rates
+    def calc_rates(self, day: datetime.date = None, drv: WeatherDataContainer = None):
+        """Delegate rate calculation to the selected evapotranspiration module."""
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
        """Delegate rate calculation to the selected evapotranspiration module.

        Args:
            day (datetime.date, optional): The current date of the simulation.
            drv (WeatherDataContainer, optional): A dictionary-like container holding
                weather data elements as key/value. The values are
                arrays or scalars. See PCSE documentation for details.
        """
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+    | IDOST | Number of days with oxygen stress  | N   | -    |
+
+    **Rate variables**
+
+    | Name  | Description                                         | Pbl | Unit      |
+    |-------|-----------------------------------------------------|-----|-----------|
+    | EVWMX | Max evaporation rate from open water surface        | Y   | cm day⁻¹  |
+    | EVSMX | Max evaporation rate from wet soil surface          | Y   | cm day⁻¹  |
+    | TRAMX | Max transpiration rate from canopy                  | Y   | cm day⁻¹  |
+    | TRA   | Actual transpiration rate from canopy               | Y   | cm day⁻¹  |
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
    | TRA   | Actual transpiration rate from canopy               | Y   | cm day⁻¹  |
    | IDOS  | Indicates oxygen stress on this day (True|False)    | N   | -         |
    | IDWS  | Indicates water stress on this day (True|False)     | N   | -         |
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+            r.TRALY = _z
+            r.RFWS = _o
+            r.RFOS = _o
+            r.RFTRA = _o
+            r.IDWS = False
+            r.IDOS = False
+            return r.TRA, r.TRAMX
+
+        kglob = 0.75 * p.KDIFTB(dvs)
+        et0_crop = torch.clamp(p.CFET * et0, min=0.0)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
        # crop specific correction on potential transpiration rate
        et0_crop = torch.clamp(p.CFET * et0, min=0.0)
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+            dsos = torch.zeros_like(dvs)
+
+        crairc = p.CRAIRC
+        sm0 = p.SM0
+        denom_ox = torch.where(crairc.abs() > self._epsilon, crairc, self._epsilon)
+        rfosmx = torch.clamp((sm0 - sm) / denom_ox, min=0.0, max=1.0)
+        rfos = rfosmx + (1.0 - torch.clamp(dsos, max=4.0) / 4.0) * (1.0 - rfosmx)
+        r.RFOS = torch.where(mask_ox, rfos, r.RFOS)
+
+        r.RFTRA = r.RFOS * r.RFWS
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
        # Transpiration rate multiplied with reduction factors for oxygen and water
        r.RFTRA = r.RFOS * r.RFWS
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 7]
```diff
@@ -0,0 +1,715 @@
...
+
+        if bool(torch.any(pre_emergence)):
+            r.EVWMX = torch.where(pre_emergence, 0.0, r.EVWMX)
+            r.EVSMX = torch.where(pre_emergence, 0.0, r.EVSMX)
+            r.TRAMX = torch.where(pre_emergence, 0.0, r.TRAMX)
+            r.TRA = torch.where(pre_emergence, 0.0, r.TRA)
+            r.TRALY = torch.where(pre_emergence, 0.0, r.TRALY)
+            r.RFWS = torch.where(pre_emergence, 1.0, r.RFWS)
+            r.RFOS = torch.where(pre_emergence, 1.0, r.RFOS)
+            r.RFTRA = torch.where(pre_emergence, 1.0, r.RFTRA)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]: the same as above, lets use a mask based on dvs.

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+        )
+        self._layer_depth_hi = torch.cumsum(thicknesses, dim=0)
+        self._layer_depth_lo = self._layer_depth_hi - thicknesses
+
+        # Internal DSOS tracker for layered oxygen-stress response
+        self._dsos = torch.zeros(self.params_shape, dtype=self.dtype, device=self.device)
+
+    def _rf_tramx_co2(self, drv: WeatherDataContainer, et0: torch.Tensor) -> torch.Tensor:
+        """Calculate CO2 reduction factor for TRAMX using CO2 from driver or parameters."""
+        p = self.params
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 9]
```diff
@@ -0,0 +1,715 @@
...
+
+    @prepare_states
+    def integrate(self, day: datetime.date = None, delt=1.0) -> None:
+        """Accumulate stress-day counters for water and oxygen stress."""
+        rfws_stress = (self.rates.RFWS < 1.0).to(dtype=self.dtype)
+        rfos_stress = (self.rates.RFOS < 1.0).to(dtype=self.dtype)
+        self.states.IDWST = self.states.IDWST + rfws_stress
+        self.states.IDOST = self.states.IDOST + rfos_stress
+
+
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]: also, I see that the function `_rf_tramx_co2` is used in other classes, we can add it here instead, something like:

```python
    def _rf_tramx_co2(self, drv: WeatherDataContainer, et0: torch.Tensor, co2_effect=True) -> torch.Tensor:
        """Calculate CO2 reduction factor for TRAMX based on atmospheric CO2 concentration."""
        if not co2_effect:
            return torch.ones_like(et0)

        if hasattr(drv, "CO2") and drv.CO2 is not None:
            co2 = _get_drv(drv.CO2, self.params_shape, dtype=self.dtype, device=self.device)
        else:
            co2 = self.params.CO2
        return self.params.CO2TRATB(co2)
```

Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/evapotranspiration.py]: nice! :100:

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 10]
```diff
@@ -0,0 +1,800 @@
...
+    | DEPNR    | Dependency number for crop sensitivity to soil moisture stress | SCr  | -    |
+    | KDIFTB   | Extinction coefficient for diffuse visible light vs DVS | TCr  | -    |
+    | IAIRDU   | Switch airducts on (1) or off (0)                       | SCr  | -    |
+    | IOX      | Switch oxygen stress on (1) or off (0)                  | SCr  | -    |
+    | CRAIRC   | Critical air content for root aeration                  | SSo  | -    |
+    | SM0      | Soil porosity                                           | SSo  | -    |
+    | SMW      | Volumetric soil moisture at wilting point               | SSo  | -    |
+    | SMFCF    | Volumetric soil moisture at field capacity              | SSo  | -    |
+    | CO2      | Atmospheric CO2 concentration (used if not in drivers)  | SCr  | ppm  |
+    | CO2TRATB | Reduction factor for TRAMX as function of CO2           | TCr  | -    |
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
    | CFET     | Correction factor for potential transpiration rate      | S  | -    |
    | DEPNR    | Dependency number for crop sensitivity to soil moisture stress | S  | -    |
    | KDIFTB   | Extinction coefficient for diffuse visible light vs DVS | T  | -    |
    | IAIRDU   | Switch airducts on (1) or off (0)                       | S  | -    |
    | IOX      | Switch oxygen stress on (1) or off (0)                  | S  | -    |
    | CRAIRC   | Critical air content for root aeration                  | S  | -    |
    | SM0      | Soil porosity                                           | S  | -    |
    | SMW      | Volumetric soil moisture at wilting point               | S  | -    |
    | SMFCF    | Volumetric soil moisture at field capacity              | S  | -    |
    | CO2      | Atmospheric CO2 concentration (used if not in drivers)  | S  | ppm  |
    | CO2TRATB | Reduction factor for TRAMX as function of CO2           | T  | -    |
```

File: src/diffwofost/physical_models/crop/evapotranspiration.py [hunk 11]
```diff
@@ -0,0 +1,800 @@
...
+    | Name  | Description                                         | Pbl | Unit      |
+    |-------|-----------------------------------------------------|-----|-----------|
+    | EVWMX | Max evaporation rate from open water surface        | Y   | cm day⁻¹  |
+    | EVSMX | Max evaporation rate from wet soil surface          | Y   | cm day⁻¹  |
+    | TRAMX | Max transpiration rate from canopy (CO2-adjusted)   | Y   | cm day⁻¹  |
+    | TRA   | Actual transpiration rate from canopy               | Y   | cm day⁻¹  |
+    | IDOS  | Indicates oxygen stress on this day (True|False)    | N   | -         |
+    | IDWS  | Indicates water stress on this day (True|False)     | N   | -         |
+    | RFWS  | Reduction factor for water stress                   | N   | -         |
+    | RFOS  | Reduction factor for oxygen stress                  | N   | -         |
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/evapotranspiration.py]:
```suggestion
    | RFWS  | Reduction factor for water stress                   | Y   | -         |
    | RFOS  | Reduction factor for oxygen stress                  | Y  | -         |
```
